### PR TITLE
ConvFrozenScaleBiasReLU fusion

### PR DESCRIPTION
**Pull request by minitu**
This PR adds ConvFrozenScaleBiasReLU fusion, built on cuDNN runtime fusion for patterns such as 'conv -> frozen batchnorm -> relu'.

### CODE REVIEW

File: apex/contrib/test/conv_bias_relu/test_conv_bias_relu.py [hunk 1]
```diff
@@ -95,11 +104,25 @@ def test_conv_bias_mask_relu(self):
...
+            loss = (out.float()**2).sum() / out.numel()
+        loss.backward()
+        with torch.cuda.amp.autocast(dtype=torch.half):
+            out_ = F.relu(self.conv2_(self.x_) * self.scale_ + self.bias_)
+            loss_ = (out_**2).sum() / out_.numel()
+        loss_.backward()
+
+        self.assertTrue(torch.allclose(out, out_, atol=2.5e-3, rtol=2.5e-3, equal_nan=True))
+        self.assertTrue(torch.allclose(self.conv2_.weight.grad, self.conv2.weight.grad, atol=1e-3, rtol=1e-3, equal_nan=True))
+        self.assertTrue(torch.allclose(self.x_.grad, self.x.grad, atol=1e-3, rtol=1e-3, equal_nan=True))
```
[the remark this answers is not in the document]
Comment (minitu) [reply apex/contrib/test/conv_bias_relu/test_conv_bias_relu.py:124]: Sure, just tested with `torch.testing.assert_close` and it seems to work fine.
The information that is provided when the assertion fails is also very useful.
Will push a change to update these.

File: apex/contrib/csrc/conv_bias_relu/conv_bias_relu.cpp [hunk 2]
```diff
@@ -1339,6 +1704,153 @@ std::vector<at::Tensor> conv_bias_mask_relu_forward(std::vector<at::Tensor> inpu
 }
 
 
+std::vector<at::Tensor> conv_cscale_cbias_relu_forward(std::vector<at::Tensor> inputs, int64_t padding, int64_t stride) {
```
[the remark this answers is not in the document]
Comment (minitu) [reply apex/contrib/csrc/conv_bias_relu/conv_bias_relu.cpp:1707]: Good catch, it only has 1 tensor here so I think for performance's sake we should return the output tensor by itself and avoid creating a `std::vector`.
I guess Gil was following the convention of returning vectors from the other fusion codes.
Let me get back to you after changing this and checking if the testbench and RetinaNet training works fine.

[the remark this answers is not in the document]
Comment (minitu) [reply apex/contrib/csrc/conv_bias_relu/conv_bias_relu.cpp:1707]: It seems this change should be applied to all the other fused forward calls that have only one output tensor as well (e.g., `conv_bias_mask_relu_forward`), although that probably should be a separate PR.

File: apex/contrib/csrc/conv_bias_relu/conv_bias_relu.cpp [hunk 3]
```diff
@@ -1339,6 +1704,149 @@ std::vector<at::Tensor> conv_bias_mask_relu_forward(std::vector<at::Tensor> inpu
...
+                             conv_stride,
+                             conv_dilation,
+                             CUDNN_DATA_HALF,
+                             x,
+                             w,
+                             s,
+                             b,
+                             y);
+
+  DEBUG_MSG("[DEBUG] conv-cscale-cbias-relu : " << y.to(at::kFloat).sum().item<float>());
```
Comment (crcrpar) [inline apex/contrib/csrc/conv_bias_relu/conv_bias_relu.cpp:1757]: wouldn't this incur host-device sync?

Comment (crcrpar) [reply apex/contrib/csrc/conv_bias_relu/conv_bias_relu.cpp:1757]: ah, this looks fine as this macro gets noop unless `DEBUG` is set

**Review by crcrpar — APPROVED** [review]

**Comment by minitu** [comment]
Confirmed that the fusion still works fine with the new commits, please merge when possible.
